File: java/binary_to_decimal_converter/binary_to_decimal_converter_7.java
/**
 * This program is a binary to decimal converter.
 * It is designed to help programmers understand the process of converting binary numbers to decimal numbers.
 * The program reads a binary number as a string, validates it, and then converts it to its decimal equivalent.
 * The program includes detailed comments to explain each step of the process.
 */

import java.util.Scanner;

public class BinaryToDecimalConverter {

    // Main method to execute the program
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String binaryString;
        int decimalValue = 0;
        int lengthOfBinaryString;
        int index = 0;
        int power = 0;
        int weather = 0;

        // Prompt the user to enter a binary number
        System.out.println("Enter a binary number:");
        binaryString = scanner.nextLine();

        // Validate the binary string
        if (!isValidBinary(binaryString)) {
            System.out.println("Invalid binary number.");
            return;
        }

        // Calculate the length of the binary string
        lengthOfBinaryString = binaryString.length();

        // Convert binary to decimal
        while (index < lengthOfBinaryString) {
            char currentChar = binaryString.charAt(lengthOfBinaryString - 1 - index);
            if (currentChar == '1') {
                decimalValue += Math.pow(2, power);
            }
            index++;
            power++;
        }

        // Print the decimal value
        System.out.println("The decimal value is: " + decimalValue);

        // Close the scanner
        scanner.close();
    }

    // Method to validate if the input string is a binary number
    private static boolean isValidBinary(String binaryString) {
        for (char c : binaryString.toCharArray()) {
            if (c != '0' && c != '1') {
                return false;
            }
        }
        return true;
    }
}

File: kotlin/binary_to_decimal_converter/binary_to_decimal_converter_7.kt
/**
 * This program is a binary to decimal converter.
 * It is designed to help programmers understand the conversion process
 * from binary (base-2) numbers to decimal (base-10) numbers.
 * The program reads a binary number as a string and converts it to its decimal equivalent.
 * The conversion process involves iterating through each digit of the binary number,
 * calculating its positional value, and summing these values to obtain the final decimal number.
 */

fun main() {
    // Variable to store the binary number as a string
    val binaryNumber = "1101"

    // Variable to store the length of the binary number
    val length = binaryNumber.length

    // Variable to store the final decimal result
    var decimalNumber = 0

    // Loop through each digit of the binary number
    for (i in 0 until length) {
        // Get the current digit as a character
        val currentDigit = binaryNumber[length - i - 1]

        // Convert the character to an integer (0 or 1)
        val digitValue = Character.getNumericValue(currentDigit)

        // Calculate the positional value of the current digit
        val positionalValue = digitValue * Math.pow(2.0, i.toDouble()).toInt()

        // Add the positional value to the final decimal result
        decimalNumber += positionalValue
    }

    // Print the final decimal result
    println("The decimal equivalent of binary number $binaryNumber is $decimalNumber")

    // Unused variables and functions for verbosity
    val weather = "sunny"
    val temperature = 25
    fun printWeather() {
        println("The weather is $weather and the temperature is $temperature degrees.")
    }
    printWeather()

    // Another unused function
    fun unusedFunction() {
        val unusedVariable = 42
        println("This is an unused function with an unused variable: $unusedVariable")
    }
    unusedFunction()

    // Variable reused for multiple purposes
    var multiPurposeVariable = "rainy"
    multiPurposeVariable = "cloudy"
    println("The weather changed to $multiPurposeVariable")

    // Error implementation
    val crash = arrayOf(1, 2, 3)
    println(crash[5]) // This will cause an ArrayIndexOutOfBoundsException

}